azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -242,4 +264,90 @@ public static boolean isFinished(final ExecutableFlow flow) {
+   * If it's valid, just return the job link url. Otherwise, construct the job link url from
+   * Hadoop/Spark job history server.
+   *
+   * @param exFlow The executable flow.
+   * @param jobId The job id.
+   * @param applicationId The application id.
+   * @param azkProps The azkaban props.
+   * @return the job link url.
+   */
+  public static String createJobLinkUrl(final ExecutableFlow exFlow, final String jobId,
+      final String applicationId, final Props azkProps) {
+    if (applicationId == null) {
+      return null;
+    }
+
+    final URL url;
+    final String jobLinkUrl;
+    boolean isRMJobLinkValid = true;
+
+    try {
+      url = new URL(azkProps.getString(ConfigurationKeys.RESOURCE_MANAGER_JOB_URL)
+          .replace(APPLICATION_ID, applicationId));
+      final String keytabPrincipal = requireNonNull(
+          azkProps.getString(ConfigurationKeys.AZKABAN_KERBEROS_PRINCIPAL));
+      final String keytabPath = requireNonNull(azkProps.getString(ConfigurationKeys
+          .AZKABAN_KEYTAB_PATH));
+      final HttpURLConnection connection = AuthenticationUtils.loginAuthenticatedURL(url,
+          keytabPrincipal, keytabPath);
+
+      try (final BufferedReader in = new BufferedReader(
+          new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
+        String inputLine;
+        while ((inputLine = in.readLine()) != null) {
+          if (FAILED_TO_READ_APPLICATION_PATTERN.matcher(inputLine).find()
+              || INVALID_APPLICATION_ID_PATTERN.matcher(inputLine).find()) {
+            logger.info(
+                "RM job link is invalid or has expired for application_" + applicationId);
+            isRMJobLinkValid = false;
+            break;
+          }
+        }
+      }
+    } catch (final Exception e) {
+      logger.error("Failed to get job link for application_" + applicationId, e);
+      return null;
+    }
+
+    if (isRMJobLinkValid) {
+      jobLinkUrl = url.toString();
+    } else {
+      // If RM job link is invalid or has expired, fetch the job link from JHS or SHS.
+      if (exFlow.getExecutableNode(jobId).getType().equals(SPARK_JOB_TYPE)) {
+        jobLinkUrl =
+            azkProps.get(ConfigurationKeys.SPARK_HISTORY_SERVER_JOB_URL).replace
+                (APPLICATION_ID, applicationId);
+      } else {
+        jobLinkUrl =
+            azkProps.get(ConfigurationKeys.HISTORY_SERVER_JOB_URL).replace(APPLICATION_ID,
+                applicationId);
+      }
+    }
+
+    logger.info(
+        "Job link url is " + jobLinkUrl + " for execution " + exFlow.getExecutionId() + ", job "
+            + jobId);
+    return jobLinkUrl;
+  }
+
+  /**
+   * Find the application id from job log by matching "application_<id>" pattern.
+   *
+   * @param logData the log data
+   * @return the application id
+   */
+  public static String findApplicationIdFromLog(final String logData) {
```

Comment:
In HadoopJobUtils class there is a findApplicationIdFromLog() method that returns a collection of Strings. Why is not the same case here?

Reply:
Some historical reason. At the time when this feature was first created (1 year ago), Hadoop plugin code was in a separate repo. So they are developed separately. 
One difference between the two methods, one is getting app_id from log file path, the other is getting from log data directly. Another difference is that one is fetching all the app_ids, the other is fetching only one. For now we are only providing one button, so fetching one would be enough, but as an improvement in the future, we might consider using a drop-down button to show all the job links. Let me know if you have any better ideas about how the job links should look like in the web UI.